hyperlink/_url.py

```diff
@@ -1449,3 +1473,272 @@ def remove(self, name):
+                fragment=_UNSET, port=_UNSET, rooted=_UNSET, userinfo=_UNSET,
+                uses_netloc=_UNSET):
+        """This replace differs a little from URL.replace. For instance, it
+        accepts userinfo as a tuple, not as a string. As with the rest
+        of the methods on DecodedURL, if you pass a reserved
+        character, it will be automatically encoded instead of an
+        error being raised.
+        """
+        if path is not _UNSET:
+            path = [_encode_reserved(p) for p in path]
+        if query is not _UNSET:
+            query = [[_encode_reserved(x)
+                      if x is not None else None
+                      for x in (k, v)]
+                     for k, v in iter_pairs(query)]
+        if userinfo is not _UNSET:
+            if not len(userinfo) == 2:
+                raise ValueError('userinfo expected sequence of'
+                                 ' ["user", "password"], got %r' % userinfo)
+            userinfo = u':'.join([_encode_reserved(p) for p in userinfo])
+        new_url = self._url.replace(scheme=scheme,
+                                    host=host,
+                                    path=path,
+                                    query=query,
+                                    fragment=fragment,
+                                    port=port,
+                                    rooted=rooted,
+                                    userinfo=userinfo,
+                                    uses_netloc=uses_netloc)
+        return type(self)(url=new_url)
+
+    def get(self, name):
+        # TODO: another reason to do this in the __init__
+        return [v for (k, v) in self.query if name == k]
+
+    def add(self, name, value=None):
+        return self.replace(query=self.query + ((name, value),))
+
+    def set(self, name, value=None):
+        query = self.query
+        q = [(k, v) for (k, v) in query if k != name]
+        idx = next((i for (i, (k, v)) in enumerate(query) if k == name), -1)
+        q[idx:idx] = [(name, value)]
+        return self.replace(query=q)
+
+    def remove(self, name):
+        return self.replace(query=((k, v) for (k, v) in self.query
+                                   if k != name))
+
+    def __repr__(self):
+        cn = self.__class__.__name__
+        return '%s(url=%r)' % (cn, self._url)
+
+    def __str__(self):
+        # TODO: the underlying URL's __str__ needs to change to make
+        # this work as the URL
+        return str(self._url)
+
+    def __eq__(self, other):
+        if not isinstance(other, self.__class__):
+            return NotImplemented
+        return self.normalize().to_uri() == other.normalize().to_uri()
+
+    def __ne__(self, other):
+        if not isinstance(other, self.__class__):
+            return NotImplemented
+        return not self.__eq__(other)
+
+    def __hash__(self):
+        return hash((self.__class__, self.scheme, self.userinfo, self.host,
```

Comment:
minor: Since equality is delegated upwards to `URL`, would it be better to do the same for hashing?

e.g.

```python
def __hash__(self):
    return hash(self.normalize().to_uri())
```

Reply:
That's a good point. I may have subtly changed my mind midway through development, and now I lean more toward not delegating up to URL for equality. I'll get that fixed, thanks!

Reply:
Whenever you're delegating `hash` like this, remember you should *also* include a tweak, so that the `DecodedURL` and the `URL` representing the same data don't hash the same.

Reply:
@glyph why is that? (correctness shouldn't be effected since dictionaries compare by equality, and I'm not entirely sure what the performance problem you're trying to forestall)

Reply:
I share @moshez's curiosity.

@mahmoud - will you follow @alexwlchan's advice about delegating `__hash__`?